Display Features support (Foldable and Cutout)

andreidiaconu opened the pull request:
Adds support for foldable devices and display cutouts, in the form of Display Features.

A display feature is an area of the display that is obstructed by a hardware feature. For now, this works only on Android. There are three types of display features:

- Hinge - a physical separator between the two displays of the device.
![display-feature-hinge](https://user-images.githubusercontent.com/1402046/109837817-3dbcaa80-7c4e-11eb-9c0c-1844ba28e94b.png)
- Fold - You can view the fold like a hinge that has zero width. It tells us where the flexible display has a crease.
![display-feature-fold](https://user-images.githubusercontent.com/1402046/109837877-4b723000-7c4e-11eb-9659-fed746079ce8.png)
- Cutout - A cutout sits at the edge of the display and usually houses camera systems.
![display-feature-cutout](https://user-images.githubusercontent.com/1402046/109837892-4f9e4d80-7c4e-11eb-911d-db9433e7727d.png)

Here is a preview of what this PR enables, on a dual-screen emulator:

<img width="1491" alt="hinge-aware-dialog-right" src="https://user-images.githubusercontent.com/1402046/109838243-a7d54f80-7c4e-11eb-9537-949c1c3c080a.png">


Issues that will be closed by this PR: 

- https://github.com/flutter/flutter/issues/49411

## Work in progress:

- Tests are still in progress
- cipd setup and DEPS update for AndroidX WindowManager library, which this PR is based on.

xster commented:
@GaryQian can you give it a look?

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  final DisplayFeatureState state;
+
+  @override
+  bool operator ==(Object other) {
```
Nit, this has typically been implemented as something like:
```
     if (identical(this, other))
        return true;
     if (other.runtimeType != runtimeType)
        return false;
     return other is DisplayFeature ...
```
Nothing wrong with your approach, but prefer consistency with rest of codebase

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
   }
 }
 
+/// Area of the display that may be obstructed by a hardware feature.This is
```
Nit, space after period.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+/// populated only on Android.
+///
+/// The [bounds] are measured in logical pixels. On devices with two screens the
+/// coordinate system starts with [0,0] in the top-left corner of the left screen
```
May want to clarify orientation of the device, as left/right can also be considered top/bottom

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+///
+/// The [bounds] are measured in logical pixels. On devices with two screens the
+/// coordinate system starts with [0,0] in the top-left corner of the left screen
+/// and expands to include the visual space between the screens and the right screen.
```
`the visual space between the left and right screens.`

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+/// The [bounds] are measured in logical pixels. On devices with two screens the
+/// coordinate system starts with [0,0] in the top-left corner of the left screen
+/// and expands to include the visual space between the screens and the right screen.
+/// The [type] can be used to determine if this display feature obstructs the screen or not.
```
New paragraph for new concept.

GaryQian replied:
Also, this can probably be better worded, the primary purpose of type doesn't seem to be whether it obstructs or not, rather it seems to express the shape and nature of the feature.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+
+  /// Area of the view occupied by this display feature, measured in logical pixels.
+  ///
+  /// On devices with two screens, the Flutter view spans from the top-right corner
```
similar comment with left/right vs top/down

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+    required this.state,
+  });
+
+  /// Area of the view occupied by this display feature, measured in logical pixels.
```
Specify `flutter view` rather than just view to sync better with next paragraph.

Beginning with `The` makes this more readable as well I think. (optional)

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  ///
+  /// For example, on a dual screen device in portrait mode:
+  ///
+  /// - [bounds.left] gives you the size of left screen, in logical pixels.
```
Use `  ///  * [bounds.left]: blah` to create a formatted bulleted list

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  /// Type of display feature, e.g. hinge, fold, cutout.
+  final DisplayFeatureType type;
+
+  /// Posture of display feature, which is populated only for folds and hinges.
```
This concept, especially with the randomly dropped terminology "posture" should have more explanation to avoid confusion if this is the entrypoint for developers reading the docs on this.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  final DisplayFeatureType type;
+
+  /// Posture of display feature, which is populated only for folds and hinges.
+  /// For cutouts, this is [DisplayFeatureState.unknown]
```
Sentences after the first intro sentence should be in separate paragraph separated by empty line

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  }
+}
+
+/// Type of display feature (screen obstruction). This enum contains both
```
First paragraph should be a single sentence that provides general overview. The parenthesized `(screen obstruction)` probably deserves a proper explanation rather than just throw in parentheses. Link `[]` to DisplayFeature class above.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+/// Type of display feature (screen obstruction). This enum contains both
+/// foldable obstructions and cutout obstructions.
+///
+/// Some, like [DisplayFeatureType.fold], can be reported without actually
```
`Some` what?

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+/// display is bent or has a crease. The [DisplayFeature] bounds can be
+/// 0-width in such cases.
+enum DisplayFeatureType {
+  unknown,
```
All public members/enums should be documented.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+/// 0-width in such cases.
+enum DisplayFeatureType {
+  unknown,
+  /// A fold in the flexible screen without a physical gap. The bounds for this
```
Here and elsewhere, first paragraph is one sentence only. Anything else should simply be put in a new paragraph.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  hinge,
+  /// A non-functional area of the screen, usually housing cameras or sensors.
+  cutout,
+}
```
Do things like in/under-screen fingerprint readers need a category? If a fingerprint icon comes up, we should probably prevent the app from reading gestures while fingerprint is being read.

GaryQian replied:
Specifically, this kind of feature would be valid to draw under, but not valid to interact with.

andreidiaconu replied:
Fingerprint readers are not included in the Display Feature API in Android. For security reasons, I don't expect this information to be exposed to developers - but as things are right now, this is not something we're looking at now.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+
+/// State of the display feature.
+///
+///   * For [DisplayFeatureType.fold]s & [DisplayFeatureType.hinge]s, the state is
```
one less space before the `*`

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  cutout,
+}
+
+/// State of the display feature.
```
This needs a bit more detail, it currently does not provide much useful info by itself that the name of the enum doesn't already provide.

GaryQian commented on lib/ui/window.dart:
```diff
@@ -225,6 +225,32 @@ abstract class FlutterView {
   ///   applications.
   WindowPadding get padding => viewConfiguration.padding;
 
+  /// Areas of the display that are obstructed by hardware features. This list is
```
First paragraph should be one sentence summary

GaryQian commented on lib/ui/window.dart:
```diff
@@ -225,6 +225,32 @@ abstract class FlutterView {
+  /// is empty.
+  ///
+  /// The space in which the [DisplayFeature.bounds] are defined includes all screens
+  /// and the space between them. For a dual-screen device, this means that the space
```
We should probably not limit to dual screen. Who knows what the future could bring :)

GaryQian commented on lib/ui/window.dart:
```diff
@@ -225,6 +225,32 @@ abstract class FlutterView {
+  /// populated only on Android. If the device has no display features, this list
+  /// is empty.
+  ///
+  /// The space in which the [DisplayFeature.bounds] are defined includes all screens
```
This sentence is a bit confusing, not clear what it is implying. The word 'space' seems to be used in multiple contexts here.

GaryQian commented on lib/ui/window.dart:
```diff
@@ -225,6 +225,32 @@ abstract class FlutterView {
+  /// [DisplayFeature.type] can be used to determine if this display feature
+  /// obstructs the screen or not. For example, [DisplayFeatureType.hinge] and
+  /// [DisplayFeatureType.cutout] both obstruct the display, while
+  /// [DisplayFeatureType.fold] is more like a crease in the display.
```
Remove `more like`

GaryQian commented on shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java:
```diff
@@ -522,7 +522,10 @@ public void setViewportMetrics(
       int systemGestureInsetBottom,
-      int systemGestureInsetLeft) {
+      int systemGestureInsetLeft,
+      int[] displayFeaturesBounds,
```
Just checking, you seem to be handling bounds as double previously, and int here (and in tests). Is this intended?

andreidiaconu replied:
Yes, int is used in Android and C++ layers, as they represent physical pixels. These get converted to logical pixels in dart. This is true for the screen size and other measurements (lines above this line).

GaryQian replied:
Thanks for the explanation!

GaryQian replied:
Optional, but it would be nice to make note of this somewhere briefly to make this more readable in the future.

GaryQian commented on shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:
```diff
@@ -360,5 +382,104 @@ public void dispatchSemanticsAction(
+   * Area of the viewport obstructed by a feature, eg. a hinge, a fold crease or camera cutout
+   *
+   * <p>Similar to {@link androidx.window.DisplayFeature}, used for reporting Fold / Hinge areas.
+   * The reason we define our own model is that we also want to support cutouts.
```
Pitch this as a more comprehensive class that handles hinges/folds/cutouts rather than a justification for existence.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+  const DisplayFeature({
+    required this.bounds,
+    required this.type,
+    required this.state,
```
We should assert that the state is unknown for cutouts to prevent an invalid DisplayFeature from being constructed.

GaryQian commented on shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:
```diff
@@ -360,5 +382,104 @@ public void dispatchSemanticsAction(
+   * <p>For foldables, the state is the posture. For cutouts, this is {@link UNKNOWN}
+   */
+  public enum DisplayFeatureState {
+    /** The display feature is a cutout or this state is new and not yet known to Flutter. */
```
Is it a good idea to combine the "new and not yet known to flutter" with the "feature is a cutout" since the cutout one is invalid while the other is unknown.

andreidiaconu replied:
I see your point. I am ok with this, since checking if a display feature is a cutout or not is meant to be checked through `type`, not `state`. The comment about cutouts here is meant as extra information about how `state` is populated.

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,100 @@ class WindowPadding {
+/// impeding drawing on the screen. They are useful for knowing where the
+/// display is bent or has a crease. The [DisplayFeature] bounds can be
+/// 0-width in such cases.
+enum DisplayFeatureType {
```
It would be great if you could include the pictures you have in the PR description here as well. You can upload assets to https://github.com/flutter/assets-for-api-docs and link/embed them in this PR.

andreidiaconu replied:
I opened PR https://github.com/flutter/assets-for-api-docs/pull/137 in order to make this happen.

GaryQian replied:
Merged! Thanks!

chinmaygarde commented:
@GaryQian Can you re-review this please? All you comments seem to have been addressed.

@andreidiaconu The presubmits are failing because of issues with the use of certain packages. Can you investigate fixing those so we can unblock landing this after Gary's review? Thanks!

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1091,6 +1130,24 @@ class ViewConfiguration {
+  /// This list is populated only on Android. If the device has no display
+  /// features, this list is empty.
+  ///
+  /// The area in which the [DisplayFeature.bounds] are defined includes all screens
```
`area` -> `coordinate space` would be more clear?

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,134 @@ class WindowPadding {
+/// coordinate system starts with [0,0] in the top-left corner of the left or top screen
+/// and expands to include both screens and the visual space between them.
+///
+/// The [type] describes the behaviour and how much the [DisplayFeature] obstructs the display.
```
`how much` -> `if`?

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,134 @@ class WindowPadding {
+}
+
+/// Type of [DisplayFeature], describing the [DisplayFeature] behaviour and how
+/// much the it obstructs the display.
```
remove how much

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,134 @@ class WindowPadding {
+/// for knowing where the display is bent or has a crease. The
+/// [DisplayFeature.bounds] can be 0-width in such cases.
+///
+/// The shape of the display feature for types [DisplayFeatureType.fold] and
```
"shape" may not be clear, you could make this clearer by mentioning one of the postures as an example

GaryQian commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1371,134 @@ class WindowPadding {
+///   the posture.
+/// * For [DisplayFeatureType.cutout]s, the state is not used and has the
+/// [DisplayFeatureState.unknown] value.
+enum DisplayFeatureState {
```
Would it make sense to also distinguish between an angle >180degrees, aka between flat and flipped? Not sure if this a concept that exists or not, just curious.

andreidiaconu replied:
This concept does not exist at this moment in WindowManager.

GaryQian commented on lib/ui/window.dart:
```diff
@@ -225,6 +225,32 @@ abstract class FlutterView {
+  /// This list is populated only on Android. If the device has no display
+  /// features, this list is empty.
+  ///
+  /// The area in which the [DisplayFeature.bounds] are defined includes all screens
```
same comment as above

GaryQian commented on lib/ui/window.dart:
```diff
@@ -225,6 +225,32 @@ abstract class FlutterView {
   ///   applications.
   WindowPadding get padding => viewConfiguration.padding;
 
+  /// Areas of the display that are obstructed by hardware features.
```
Since this seems to be repeated docs, you may want to use the templates system to avoid repeating documentation that is the same in multiple places

GaryQian commented on lib/web_ui/lib/src/ui/platform_dispatcher.dart:
```diff
@@ -270,6 +274,49 @@ abstract class WindowPadding {
+  final DisplayFeatureState state;
+
+  @override
+  bool operator ==(Object other) {
```
Sync up this implementation with non-web version

GaryQian commented:
Most of the implementation lgtm with minor nits, still needs tests to land.

andreidiaconu commented:
@chinmaygarde The package issues stem from the fact that this needs a new jar added to the CIPD bundle. I don't have `WRITER` access yet in order to solve this. I've written to Xiao Yu for help with permissions.

@GaryQian Tests will be updated shortly (1-2 days).

xster commented on shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java:
```diff
@@ -271,7 +274,23 @@ public void setViewportMetrics(@NonNull ViewportMetrics viewportMetrics) {
-            + viewportMetrics.viewInsetBottom);
+            + viewportMetrics.systemGestureInsetRight
+            + "\n"
+            + "Display Features Count: "
```
Be consistent with the wording

xster reviewed (commented):
This is looking pretty good! Thanks for iterating through this with Gary. 

I think this PR is ready. My only remaining concern is the alpha package from androidx.window.

andreidiaconu commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1091,6 +1130,27 @@ class ViewConfiguration {
+  /// This list is populated only on Android. If the device has no display
+  /// features, this list is empty.
+  ///
+  /// The coordinate space in which the [DisplayFeature.bounds] are defined includes all screens
```
Ok, tried to improve it by saying "The coordinate space [...] spans across the screens currently in use".

andreidiaconu commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1374,136 @@ class WindowPadding {
+/// for foldable features.
+///
+/// The posture is the shape made by the parts of the flexible screen or
+/// physical screen panels. Postures correspond to values found in
```
Ok, made it clear that the 2 concepts are similar but not tied.

andreidiaconu commented on lib/ui/platform_dispatcher.dart:
```diff
@@ -1314,6 +1374,136 @@ class WindowPadding {
+  postureFlat,
+  /// Fold angle is in an intermediate position between opened and closed state.
+  ///
+  /// There is a non-flat angle between parts of the flexible screen or between physical screen panels.
```
Talking about angle numbers can also be confusing, as 0 and 180 could mean the same thing to two different people. Tried to make it clearer: "There is a non-flat angle [...] such that the screens start to face each other."

andreidiaconu commented on shell/platform/android/io/flutter/embedding/android/FlutterView.java:
```diff
@@ -421,6 +442,95 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
+  @Override
+  protected void onAttachedToWindow() {
+    super.onAttachedToWindow();
+    windowManager.registerLayoutChangeCallback(
```
Good idea. I did, however had to create the object in separate method, so I could mock that during testing.

andreidiaconu commented:
@GaryQian @chinmaygarde Finished updating the tests. Thank you for the help so far.
@xster Would a beta version of AndroidX WindowManager be stable enough to land this, considering beta means no major api changes?

xster commented:
Thanks for all the work so far @andreidiaconu! A beta maven package would sound good to me!

chinmaygarde commented:
This seems good to go barring the presubmits. Any updates? Lets land this ASAP.

xster commented:
@chinmaygarde we're waiting for a more stable version of the androidx library we're depending on to be released.

chinmaygarde commented:
Marking this as WIP till the AndroidX dependency situation is resolved. This way, we won't trip over this again in triage. Please land when able.

goderbauer commented:
Just wondering what the status about this is. Did the AndroidX dependency get resolved?

andreidiaconu commented:
It is currently in alpha06, getting closer to beta.

On Wed, 2 Jun 2021 at 21:25, Michael Goderbauer ***@***.***>
wrote:

> Just wondering what the status about this is. Did the AndroidX dependency
> get resolved?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/engine/pull/24756#issuecomment-853281001>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKWJPSKW6YRKKHV2UO2H2DTQZZP7ANCNFSM4YRRSZFA>
> .
>

andreidiaconu commented:
Migrated to WindowManager alpha08, which we expect is very close to beta in terms of API.

Sunbreak commented:
> Marking this as WIP till the AndroidX dependency situation is resolved. This way, we won't trip over this again in triage. Please land when able.

[Jetpack WindowManager](https://developer.android.com/jetpack/androidx/releases/window) is now in beta and the Android team publish an article for the release: https://medium.com/androiddevelopers/unbundling-the-windowmanager-fa060adb3ce9

Is it stable enough?

andreidiaconu commented:
Yes, it should be. I am working on bringing this branch up to latest master
and latest WindowManager beta.

On Fri, 20 Aug 2021 at 08:39 Sunbreak ***@***.***> wrote:

> Marking this as WIP till the AndroidX dependency situation is resolved.
> This way, we won't trip over this again in triage. Please land when able.
>
> Jetpack WindowManager
> <https://developer.android.com/jetpack/androidx/releases/window> is now
> in beta and the Android team publish an article for the release:
> https://medium.com/androiddevelopers/unbundling-the-windowmanager-fa060adb3ce9
>
> Is it stable enough?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/engine/pull/24756#issuecomment-902445443>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKWJPXU4FUI2WKA42Y3JPTT5XTBDANCNFSM4YRRSZFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

xster commented:
Woohoo, the long wait is over :)

andreidiaconu commented:
I updated to beta01 `androidx.window`. Can we now remove the WIP tag? 🎉

Also, I updated [android_embedding_bundle/build.gradle](https://github.com/flutter/engine/blob/0e6fe2a0cc38172335e9e1e7ea6b27d55808bf8e/tools/cipd/android_embedding_bundle/build.gradle) and wanted to proceed with `cipd create` but I lack permissions. Without this step, I see that ci fails. Can I get cipd permissions or is there a way to get an updated version?  Thank you - @xster @chinmaygarde

blasten commented:
@chinmaygarde could you take another look when you get a chance?

chinmaygarde commented:
I think this looks fine after Gary's review. There is a conflict but we should land this once this is resolved.

andreidiaconu commented:
> I think this looks fine after Gary's review. There is a conflict but we should land this once this is resolved.

Resolved the conflict.

andreidiaconu commented:
`build_and_test_linux_unopt_debug` failed and I'm not sure why or what I can do about it. How does one `deflake`?

andreidiaconu commented:
Triggered another ci run

andreidiaconu commented:
@chinmaygarde The tree is green and this can land now, the tag you added was removed automatically due to a flaky build.

GaryQian commented:
FYI @andreidiaconu Looks like this was reverted due to failures rolling this into flutter framework: https://github.com/flutter/flutter/pull/89430#issuecomment-912514713

andreidiaconu commented:
I know and am currently working on fixing it in a separate branch of the
framework repo. Most of the ci steps fail due to gradle dependency lock
files. Might get it solved over the weekend and will update you when we can
try to land it again.

If you have any guidance please let me know. Thank you!

On Fri, 3 Sep 2021 at 22:30 Gary Qian ***@***.***> wrote:

> FYI @andreidiaconu <https://github.com/andreidiaconu> Looks like this was
> reverted due to failures rolling this into flutter framework: flutter/flutter#89430
> (comment)
> <https://github.com/flutter/flutter/pull/89430#issuecomment-912514713>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/engine/pull/24756#issuecomment-912763239>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKWJPQMJN4NTDL3FW7V7R3UAEO3RANCNFSM4YRRSZFA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
